bundles/alpha.codegen/src/alpha/codegen/writeC/SystemConverter.xtend

```diff
@@ -0,0 +1,472 @@
+import alpha.codegen.isl.PolynomialConverter
+import alpha.model.AlphaSystem
+import alpha.model.StandardEquation
+import alpha.model.UseEquation
+import alpha.model.Variable
+import alpha.model.transformation.Normalize
+import alpha.model.transformation.StandardizeNames
+import fr.irisa.cairn.jnimap.isl.ISLPWQPolynomial
+import java.util.ArrayList
+
+import static extension alpha.codegen.Factory.customExpr
+import static extension alpha.codegen.isl.PolynomialConverter.convert
+import static extension alpha.codegen.writeC.Common.getEvalName
+import static extension alpha.codegen.writeC.Common.getFlagName
+import static extension alpha.model.util.AlphaUtil.copyAE
+import static extension alpha.model.util.CommonExtensions.toArrayList
+import static extension fr.irisa.cairn.jnimap.barvinok.BarvinokBindings.card
+
+/** Converts an Alpha system to the simplified C AST. */
+class SystemConverter {
+	/** The list of all variables allocated by the entry point function which need to be freed before returning. */
+	protected val ArrayList<String> allocatedVariables
+
+	/**
+	 * A counter for the number of statement macros that have been created.
+	 * This is used for determining the names said macros.
+	 */
+	var nextStatementId = 0
+	
+	/**
+	 * Flag to indicate that the C code being output should be compatible
+	 * with the wrapper produced by the older AlphaZ.
+	 */
+	protected val boolean oldAlphaZCompatible
+	
+	/** The instance of the program builder. */
+	protected val ProgramBuilder program
+	
+	/** The system being converted into a C program. */
+	protected val AlphaSystem system
+	
+	/** Protected constructor. */
+	protected new(AlphaSystem system, boolean oldAlphaZCompatible) {
+		this.allocatedVariables = newArrayList
+		this.oldAlphaZCompatible = oldAlphaZCompatible
+		this.program = ProgramBuilder.start
+		this.system = system
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 */
+	def static convert(AlphaSystem system) {
+		return convert(system, false)
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 * 
+	 * If requested, the code produced will aim for compatibility with
+	 * the older version of AlphaZ (although with no guarantees).
+	 * This means that the inputs to the entry point will assume
+	 * that memory for the system's inputs and outputs were allocated
+	 * via the bounding box method, as opposed to linearized memory.
+	 * 
+	 * Note: since the newer Alpha language doesn't have typing,
+	 * all indices are assumed to be of type "long",
+	 * and all data values are of type "float".
+	 */
+	def static convert(AlphaSystem system, boolean oldAlphaZCompatible) {
+		// Duplicate the system and then preprocess the duplicate
+		// by applying normalization and name standardization.
+		val duplicate = system.copyAE
+		Normalize.apply(duplicate)
+		StandardizeNames.apply(duplicate)
+		
```

Comment:
Will this handle reductions at arbitrary locations in the AST? In other words, does NormalizeReduction also need to be called? I think no, but it’d be good if you can confirm.

Reply:
Yes, reductions can be anywhere, even without calling `Normalize`. When a reduce expression is encountered, a new function for performing the reduction is generated, and a call to that function is inserted where the reduce expression was found.